config/config-feature-flags.yaml

```diff
@@ -35,6 +34,15 @@ data:
   # working directory if not set by the user but this will change
   # in an upcoming release.
   #
-  # See https://github.com/tektoncd/pipeline/issues/1836 for more
-  # info.
+  # See https://github.com/tektoncd/pipeline/issues/1836 for more info.
   disable-working-directory-overwrite: "false"
+
+  # Setting this flag to "true" will set the `tekton.dev/ready` annotation
+  # at pod creation time for Taskruns without sidecars.
+  #
+  # Enabling this option should decrease the time it takes for a TaskRun to
+  # start running. However, for clusters that use injected sidecars e.g. istio
+  # enabling this option can lead to unexpected behavior.
+  #
+  # See https://github.com/tektoncd/pipeline/issues/2080 for more info.
+  enable-ready-annotation-on-pod-create: "false"
```

Comment:
are we 100% sure we want to use a feature flag for this? i feel like we're quickly going to get to a point where most new features use feature flags - maybe this is what we want!

either way, I assume the idea is that in a future release we'd turn this to "true" by default, and maybe remove the flag? If so can we create an issue to track this and make sure we do it?

Reply:
The reason we need this in a feature-flag is due to us supporting injected sidecars like Istio. If the cluster uses injected sidecars, the controller can not set the Ready annotation early since the sidecar injector might inject a sidecar container that the controller does not know about during pod creation time.

If we decide that injected sidecars are something we do not support injected sidecars, then yes, we do not need to feature-flag this. Though if I recall correctly, the feature was initially designed this way because we explicitly wanted to support the injected sidecar use case (correct me if I'm wrong @sbwsg!)

Reply:
hm interesting - we definitely want to support injected sidecars!

My understanding of the term "feature flag" is that it's used to control rollouts of features that we eventually intend to make the default - does that match other ppl's understanding? If so, I think this would make sense as a command line config param maybe? or it's own config map? i.e. something separate from feature flags

Or I'm totally wrong and it's normal to use "feature flags" to refer to functionality you intend to retain over the long run.

>  If the cluster uses injected sidecars, the controller can not set the Ready annotation early since the sidecar injector might inject a sidecar container that the controller does not know about during pod creation time.

I'm a bit sad that we can't find a solution that handles both cases well. 🤔  Have we brainstormed any ideas around this? e.g. what if we did something like:
- set the annotation to ready immediately
- be notified if the pod is modified (i.e. an admission controller has injected a sidecar), set the annotation to not ready <-- is it possible to do this before the pod actually started running? i wonder how crazy it would be to have our own admission controller for this

Is it possible we could name the flag after after exactly what it's intended for? e.g. something like `running-in-environment-with-injected-sidecars`? I think it would even be reasonable to default it to false in that case eventually, i.e. apply this optimization by default

Reply:
> My understanding of the term "feature flag" is that it's used to control rollouts of features that we eventually intend to make the default - does that match other ppl's understanding? 

Will add a new `options` or `features` configMap for this.

> I'm a bit sad that we can't find a solution that handles both cases well. 🤔 Have we brainstormed any ideas around this? e.g. what if we did something like:

Some discussion in https://github.com/tektoncd/pipeline/issues/2080 and https://github.com/tektoncd/pipeline/issues/701. Short of the Sidecar KEP becoming a reality, I think this is the best we can do at this moment

>  <-- is it possible to do this before the pod actually started running?
 No way to guarantee that unfortunately.

> Is it possible we could name the flag after after exactly what it's intended for? e.g. something like running-in-environment-with-injected-sidecars? I think it would even be reasonable to default it to false in that case eventually, i.e. apply this optimization by default

I'll change the flag name for now. Maybe we can set the default to false in a later change.